lib/ld-eventsource/impl/buffered_line_reader.rb

```diff
@@ -0,0 +1,73 @@
+
+module SSE
+  module Impl
+    class BufferedLineReader
+      #
+      # Reads a series of data chunks from an enumerator, and returns an enumerator that
+      # parses/aggregates these into text lines. The line terminator may be CR, LF, or
+      # CRLF for each line; terminators are not included in the returned lines. When the
+      # input data runs out, the output enumerator ends and does not include any partially
+      # completed line.
+      #
+      # @param [Enumerator] chunks  an enumerator that will yield strings from a stream
+      # @return [Enumerator]  an enumerator that will yield one line at a time
+      #
+      def self.lines_from(chunks)
+        buffer = ""
+        position = 0
+        line_start = 0
+        last_char_was_cr = false
+
+        Enumerator.new do |gen|
+          chunks.each do |chunk|
+            buffer << chunk
```

Comment:
Buffer length may get slightly costly as the buffer grows. I don't think this is a concern up until you get above 200Mb or so, but might be worth considering a running numerical tracker if that bottlenecks in the future.

Reply:
I'm not sure what you're getting at. The buffer is only going to grow large if either 1. a large amount of data containing multiple lines arrives from the socket before we've had a chance to parse it— in which case we will throw it away as soon as we have parsed those lines, which we would have to do anyway— or 2. there is one very long line of event data, in which case we are going to need to store that data as a single string in the SSE event anyway; there would be no savings in breaking it up if we're simply going to re-concatenate it later.